internal/speakerlist/speakerlist.go

```diff
@@ -0,0 +1,301 @@
+	// changes for k8s watch. It is assumed that operations there don't
+	// block too much time (is single thread), so make it 'big' to reduce
+	// the chances of blocking.
+	// K8s max node size is 5k, and this call back on node rejoin can be
+	// called 2-3 times (node change to Not Ready, then Ready again and
+	// sometimes another change is detected too), so we make it 5k * 3
+	sl.mlJoinCh = make(chan struct{}, 5000*3)
+
+	// ChannelEventDelegate hint that it should not block, so make mlEventCh 'big'
+	sl.mlEventCh = make(chan memberlist.NodeEvent, 1024)
+	mconfig.Events = &memberlist.ChannelEventDelegate{Ch: sl.mlEventCh}
+	var err error
+	sl.mList, err = memberlist.Create(mconfig)
+	if err != nil {
+		sl.l.Log("op", "startup", "error", err, "msg", "failed to create memberlist")
+		return nil, err
+	}
+	return &sl, nil
+}
+
+// Start starts the needed goroutines
+func (sl *SpeakerList) Start(client *k8s.Client) error {
+	sl.client = client
+
+	if sl.mList == nil {
+		return nil
+	}
+
+	go sl.memberlistWatchEvents()
+	go sl.mlJoinLoop()
+
+	return nil
+}
+
+// updateMlIPs runs forever updating the sl.mlSpeakersIPs slice with the
+// IPs of all the speaker pods. As this queries the API server, it waits at
+// least 5 minutes between queries, to avoid self-induced API server DoS (should
+// be trated with care).
+func (sl *SpeakerList) speakersIPsLoop() {
+	for {
+		// Create the ticker at the start of each loop, so we wait at
+		// least the ticker interval before the ticker case is true.
+		// If we create it outside of the loop, it will tick every 5
+		// minutes and if the blocking call happens to take longer, the
+		// next tick will be really soon and we may overload the API
+		// server.
+		ticker := time.NewTicker(5 * time.Minute)
+
+		select {
+		case <-sl.stopCh:
+			return
+		case <-ticker.C:
+			// This blocks until the apiserver responds.
+			iplist, err := sl.mlSpeakers()
+			if err != nil {
+				continue
+			}
+
+			sl.mlMux.Lock()
+			sl.mlSpeakersIPs = iplist
+			sl.mlMux.Unlock()
+		}
+	}
+}
+
+func (sl *SpeakerList) mlSpeakers() ([]string, error) {
+	// This call blocks until we get a response of the API server.
+	// In the client-go version we are using, there is no way to use a
+	// context or timeout.
+	// TODO: When updating client-go, this code can be simplified by using a
+	// context in the following way:
+	// If we use a context, we can specify a timeout and there is no need to
+	// have the speakersIPsLoop() running in the background. This can just
+	// be called from mlJoin() with a context (timeout) and if doesn't
+	// answer, use sl.mlSpeakersIPs.  If it doesn't time out, just update
+	// sl.mlSpeakersIPs with the answer and continue to do the join.
+	iplist, err := sl.client.GetPodsIPs(sl.namespace, sl.labels)
+	if err != nil {
+		sl.l.Log("op", "Member Detection", "error", err, "msg", "failed to get PodsIPs")
+		return nil, err
+	}
+
+	return iplist, nil
+}
+
+func (sl *SpeakerList) mlJoinLoop() {
```

Comment:
Let's name the function based on its functionality.

Reply:
I think the name is the functionality: memberlist (ml) join loop is what this does

Is a very used pattern, like in Kubernetes code base. What name do you suggest?

Reply:
@johananl friendly ping?

Reply:
Done!

Renamed also `speakersIPsLoop()` to `mlUpdateSpeakers()`. Let me know if you think this name is good or some other name is better